### PR TITLE
Resolve raw params

### PR DESCRIPTION
**Pull request by GuyAv46**
Added ResolveParams function to VecSimIndex, including the struct VecSimRawParam that holds a parameter in <string,string> key-value fashion.
Also added tests for hnsw and br indexes ResolveParams.

### CODE REVIEW

File: src/VecSim/vec_sim_index.h [hunk 1]
```diff
@@ -48,6 +48,15 @@ class VecSimIndex : public VecsimBaseObject {
      */
     virtual size_t indexSize() const = 0;
 
+    /**
+     * @brief Resolves VecSimRawParam array and generate VecSimQueryParams struct.
+     * @param index the index whose size is requested.
+     * @param rparams array of raw params to resolve.
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_index.h:54]: you choose to take the assumption that you will get an instance of `arr.h` this should not be the basic assumption. In C API, you give a pointer and a length of the array

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp [hunk 2]
```diff
@@ -50,6 +50,30 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
 
 int HNSWIndex::deleteVector(size_t id) { return this->hnsw->removePoint(id); }
 
+int HNSWIndex::resolveParams(VecSimRawParam *rparams, VecSimQueryParams *qparams) {
+    bzero(qparams, sizeof(VecSimQueryParams));
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnswlib_c.cpp:54]: validate null

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 3]
```diff
@@ -154,6 +154,11 @@ int BruteForceIndex::deleteVector(size_t label) {
 
 size_t BruteForceIndex::indexSize() const { return this->count; }
 
+int BruteForceIndex::resolveParams(VecSimRawParam *rparams, VecSimQueryParams *qparams) {
+    bzero(qparams, sizeof(VecSimQueryParams));
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force.cpp:158]: validate NULL
also, what about the index type if you are setting to zeros?

File: tests/unit/test_bruteforce.cpp [hunk 4]
```diff
@@ -837,3 +838,28 @@ TEST_F(BruteForceTest, brute_force_batch_iterator_corner_cases) {
...
+                                        .metric = VecSimMetric_L2,
+                                        .initialCapacity = 0,
+                                        .blockSize = 5}};
+    VecSimIndex *index = VecSimIndex_New(&params);
+
+    VecSimQueryParams qparams;
+
+    VecSimRawParam *rparams = array_new<VecSimRawParam>(1);
+
+    ASSERT_TRUE(VecSimIndex_ResolveParams(index, rparams, &qparams));
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:857]: please validate qparams

File: tests/unit/test_hnswlib.cpp [hunk 5]
```diff
@@ -913,3 +914,51 @@ TEST_F(HNSWLibTest, hnsw_batch_iterator_advanced) {
...
+                                          .M = M,
+                                          .efConstruction = ef,
+                                          .efRuntime = ef}};
+    VecSimIndex *index = VecSimIndex_New(&params);
+
+    VecSimQueryParams qparams;
+
+    VecSimRawParam *rparams = array_new<VecSimRawParam>(2);
+
+    ASSERT_TRUE(VecSimIndex_ResolveParams(index, rparams, &qparams));
```
Comment (DvirDukhan) [inline tests/unit/test_hnswlib.cpp:937]: why this is true?

File: tests/unit/test_hnswlib.cpp [hunk 6]
```diff
@@ -913,3 +914,51 @@ TEST_F(HNSWLibTest, hnsw_batch_iterator_advanced) {
...
+
+    array_append(rparams, (VecSimRawParam){
+                              .name = "ef_runtime", .nameLen = 10, .value = "100", .valLen = 3});
+    ASSERT_TRUE(VecSimIndex_ResolveParams(index, rparams, &qparams));
+
+    array_append(rparams, (VecSimRawParam){
+                              .name = "ef_runtime", .nameLen = 10, .value = "100", .valLen = 3});
+    ASSERT_FALSE(VecSimIndex_ResolveParams(index, rparams, &qparams));
+
+    array_hdr(rparams)->len--;
```
Comment (DvirDukhan) [inline tests/unit/test_hnswlib.cpp:947]: array_pop? array_del_fast? don't break abstractions. If you need a function, write it

File: tests/unit/test_hnswlib.cpp [hunk 7]
```diff
@@ -913,3 +914,51 @@ TEST_F(HNSWLibTest, hnsw_batch_iterator_advanced) {
...
+
+    VecSimQueryParams qparams;
+
+    VecSimRawParam *rparams = array_new<VecSimRawParam>(2);
+
+    ASSERT_TRUE(VecSimIndex_ResolveParams(index, rparams, &qparams));
+
+    array_append(rparams, (VecSimRawParam){
+                              .name = "ef_runtime", .nameLen = 10, .value = "100", .valLen = 3});
+    ASSERT_TRUE(VecSimIndex_ResolveParams(index, rparams, &qparams));
```
Comment (DvirDukhan) [inline tests/unit/test_hnswlib.cpp:941]: validate qparams?

File: src/VecSim/algorithms/brute_force/brute_force.cpp (hunk 3 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:158]: VecSimQueryParams doesn't hold type field. Each index is looking for its part in the union.
Right now VecSimQueryParams is just size_t for hnsw.

File: tests/unit/test_hnswlib.cpp (hunk 5 above)
Comment (GuyAv46) [reply tests/unit/test_hnswlib.cpp:937]: I assume calling VecSimIndex_ResolveParams without params is OK, So empty array should return true.

File: tests/unit/test_hnswlib.cpp (hunk 7 above)
Comment (GuyAv46) [reply tests/unit/test_hnswlib.cpp:941]: the value of qparams is not defined if rparams has bad params. should I zero up qparams?

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp [hunk 8]
```diff
@@ -50,6 +50,30 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
...
+    for (int i = 0; i < array_len(rparams); i++) {
+        if (!strncasecmp(rparams[i].name, VecSimCommonStrings::HNSW_EF_RUNTIME_STRING,
+                         rparams[i].nameLen)) {
+            if (qparams->hnswRuntimeParams.efRuntime != 0) { // Already been set.
+                return false;
+            } else {
+                char *ep;
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnswlib_c.cpp:61]: Consider adding documentation for what `ep` is used for

**Review by alonre24 — COMMENTED** [review]
Looks good, added few comments. Also, consider return an error message in case that `VecSimIndex_ResolveParams` fails (@DvirDukhan ?)

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp [hunk 9]
```diff
@@ -50,6 +50,30 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
 
 int HNSWIndex::deleteVector(size_t id) { return this->hnsw->removePoint(id); }
 
+int HNSWIndex::resolveParams(VecSimRawParam *rparams, VecSimQueryParams *qparams) {
+    bzero(qparams, sizeof(VecSimQueryParams));
+    for (int i = 0; i < array_len(rparams); i++) {
+        if (!strncasecmp(rparams[i].name, VecSimCommonStrings::HNSW_EF_RUNTIME_STRING,
+                         rparams[i].nameLen)) {
+            if (qparams->hnswRuntimeParams.efRuntime != 0) { // Already been set.
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnswlib_c.cpp:58]: If someone tries to set this parameter twice.

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp [hunk 10]
```diff
@@ -50,6 +50,30 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
...
+                return false;
+            } else {
+                char *ep;
+                errno = 0;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnswlib_c.cpp:62]: IDK. maybe there was some error earlier and the value was not changed?

[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnswlib_c.cpp:65]: Then I would just check if `errno==ERANGE`. But it's not crucial...

File: src/VecSim/vec_sim_common.h [hunk 11]
```diff
@@ -26,6 +26,26 @@ typedef enum { VecSimAlgo_BF, VecSimAlgo_HNSWLIB } VecSimAlgo;
...
+    size_t valLen;
+} VecSimRawParam;
+
+typedef enum {
+    VecSimErr_OK = 0,
+    VecSimErr_MissingParamStruct,
+    VecSimErr_AlreadySet,
+    VecSimErr_UnknownParam,
+    VecSimErr_BadValue
+} VecSimResolveCode;
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_common.h:47]: This is for the specific param resolve use case, right? if so:
1. please us specific param resolve prefix like `VecSimParamResolverErr_BadValue`
2. don't use contradicting words in the same code `VecSimErr_Ok`
3. take the above error code and have it as a define, so if in the future we will need additional error code usage, we will have `VecSim_OK` shared

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 12]
```diff
@@ -154,6 +153,18 @@ int BruteForceIndex::deleteVector(size_t label) {
 
 size_t BruteForceIndex::indexSize() const { return this->count; }
 
+VecSimResolveCode BruteForceIndex::resolveParams(VecSimRawParam *rparams, int paramNum,
+                                                 VecSimQueryParams *qparams) {
+    if (!qparams) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:158]: if rparams is null I assume paramNum is 0

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp [hunk 13]
```diff
@@ -50,6 +49,34 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
...
+        if (!strncasecmp(rparams[i].name, VecSimCommonStrings::HNSW_EF_RUNTIME_STRING,
+                         rparams[i].nameLen)) {
+            if (qparams->hnswRuntimeParams.efRuntime != 0) {
+                return VecSimErr_AlreadySet;
+            } else {
+                char *ep; // for checking strtoll used all rparams[i].valLen chars.
+                errno = 0;
+                long long val = strtoll(rparams[i].value, &ep, 0);
+                if (val <= 0 || val == LLONG_MAX || errno != 0 ||
+                    (rparams[i].value + rparams[i].valLen) != ep) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnswlib_c.cpp:68]: there is no str to size_t func, so Im using str to long long. this way I can also make sure we got positive number, and also that it wasnt to large number.
the last test checks that the entire rparams[i].value was used. ep will hold the char the strtoll was stoped on, and the test tests that it is the entire word. we catch here input like "3.14", "123text" (this isnt possible with RediSearch parser) and so on.

File: tests/unit/test_hnswlib.cpp [hunk 14]
```diff
@@ -913,3 +914,66 @@ TEST_F(HNSWLibTest, hnsw_batch_iterator_advanced) {
...
+              VecSimErr_OK);
+    ASSERT_EQ(memcmp(&qparams, &zero, sizeof(VecSimQueryParams)), 0);
+
+    array_append(rparams, (VecSimRawParam){
+                              .name = "ef_runtime", .nameLen = 10, .value = "100", .valLen = 3});
+    ASSERT_EQ(VecSimIndex_ResolveParams(index, rparams, array_len(rparams), &qparams),
+              VecSimErr_OK);
+    ASSERT_EQ(qparams.hnswRuntimeParams.efRuntime, 100);
+    qparams.hnswRuntimeParams.efRuntime = 0;
+    ASSERT_EQ(memcmp(&qparams, &zero, sizeof(VecSimQueryParams)), 0);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnswlib.cpp:948]: it checks that the only field we set is `efRuntime`

File: src/VecSim/algorithms/brute_force/brute_force.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:158]: this is a wrong assumption. you can easily test it

File: src/VecSim/algorithms/hnsw/hnswlib_c.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/VecSim/algorithms/hnsw/hnswlib_c.cpp:68]: but please add it in a comment :)

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**